addons/web_editor/static/src/js/backend/field_html.js

```diff
@@ -40,14 +47,74 @@ var FieldHtml = basic_fields.DebouncedField.extend(TranslatableFieldMixin, {
     quickEditExclusion: [
         '[href]',
     ],
-
+    dynamicPlaceholderOptions: {
+        readonly: false,
+        needDefaultValue: true,
+        cancelOnEscape: true,
+        chainedTitle: true,
+        filter: (model) => !["one2many", "boolean", "many2many"].includes(model.type)
+    },
     custom_events: {
         wysiwyg_focus: '_onWysiwygFocus',
         wysiwyg_blur: '_onWysiwygBlur',
         wysiwyg_change: '_onChange',
         wysiwyg_attachment: '_onAttachmentChange',
     },
 
+    /**
+     * @override
+     */
+    init: function () {
+        this._super(...arguments);
+        // TODO this break the mass_mailing editor tour ( the content of the field html is reset )
+        //      why was this code here in the first place ?
+        // if (this.nodeOptions && this.nodeOptions.dynamic_placeholder) {
+        //     // When the dynamic placeholder is active,
+        //     // the recordData need to be updated when `mailing_model_real` change
+        //     this.resetOnAnyFieldChange = true;
+        // }
+    },
+
+    openDynamicPlaceholder: async function (baseModel, chain = []) {
+        if (!baseModel) {
+            return;
+        }
+
+        const modelSelector = new ModelFieldSelectorPopOver(this, baseModel, [], this.dynamicPlaceholderOptions);
+        modelSelector.on("field_chain_changed", undefined, (ev) => {
+            this.wysiwyg.odooEditor.editable.focus();
+            if (ev.data.chain.length) {
+                let dynamicPlaceholder = "object." + ev.data.chain.join('.');
+                const defaultValue = ev.data.defaultValue;
+                dynamicPlaceholder += defaultValue && defaultValue !== '' ? ` or '''${defaultValue}'''` : '';
+
+                const t = document.createElement('T');
+                t.setAttribute('t-out', dynamicPlaceholder);
+                this.wysiwyg.odooEditor.execCommand('insert', t.outerHTML, 'html');
+                setSelection(...rightPos(t));
+                this.wysiwyg.odooEditor.editable.focus();
+            }
+            modelSelector.destroy();
+        });
+        modelSelector.on("field_chain_cancel", undefined, () => {
+            this.wysiwyg.odooEditor.editable.focus();
+            modelSelector.destroy();
+        });
+        await modelSelector.insertAfter(this.$el);
+        modelSelector.open(chain, true);
```

Comment:
There is quite a lot of boilerplate code that is duplicated between the two `openDynamicPlaceholder` methods. Wouldn't it be possible to abstract it and only have to define what happens on `field_chain_changed` ?

Reply:
I'll look into that

Reply:
Since the comment was neither marked as resolved nor outdated, it is not easy from here to assess whether it has been addressed or not. Since other changes are still needed before merging this PR, I'm leaving this one open in the meantime.